compiler/rustc_lint/src/builtin.rs

```diff
@@ -2598,6 +2598,200 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
     }
 }
 
+////////////////////////////////
+declare_lint! {
+    /// The `mem_uninitialized` lint detects all uses of `std::mem::uninitialized`.
+    ///
+    /// This function is extremely dangerous, and nearly all uses of it cause immediate Undefined
+    /// Behavior.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(mem_uninitialized)]
+    /// fn main() {
+    ///     let x: i32 = unsafe { std::mem::uninitialized() };
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating an invalid value is undefined behavior, and nearly all types are invalid when left
+    /// uninitialized.
+    pub MEM_UNINITIALIZED,
+    Warn,
+    "use of mem::uninitialized",
+    @future_incompatible = FutureIncompatibleInfo {
+        reference: "FIXME: fill this in",
+        reason: FutureIncompatibilityReason::CustomReportNow(""),
+        explain_reason: false,
+    };
+}
+
+declare_lint_pass!(MemUninitialized => [MEM_UNINITIALIZED]);
+
+impl<'tcx> LateLintPass<'tcx> for MemUninitialized {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'_>) {
+        /// Information about why a type cannot be initialized this way.
+        /// Contains an error message and optionally a span to point at.
+        type InitError = (String, Option<Span>);
+
+        /// Determine if this expression is a "dangerous initialization".
+        fn is_dangerous_init(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+            if let hir::ExprKind::Call(ref path_expr, _) = expr.kind {
+                // Find calls to `mem::{uninitialized,zeroed}` methods.
+                if let hir::ExprKind::Path(ref qpath) = path_expr.kind {
+                    if let Some(def_id) = cx.qpath_res(qpath, path_expr.hir_id).opt_def_id() {
+                        if cx.tcx.is_diagnostic_item(sym::mem_uninitialized, def_id) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            false
+        }
+
+        /// Return `None` only if we are sure this type does
+        /// allow being left uninitialized.
+        fn ty_find_init_error<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<InitError> {
```

Comment:
Why is this so clever?

I think the idea is to just warn on *all* uses of this function. Basically this is a very strong form of deprecation. We might even do this via a new attribute rather than a specific lint -- `#[extremely_deprecated]` or so, which makes the existing deprecation lint into an FCW lint.

Reply:
Oh, looks like this was specifically requested by @Mark-Simulacrum. However my understanding from previous discussions was that we didn't want to differentiate here -- we want to slowly work towards removing that function entirely. If we had a time machine we'd erase it from history; lacking that we will remove it as much as we possibly can.

Reply:
Yeah, I'm not sure if we agreed on exactly *when* it should lint. https://github.com/rust-lang/rust/pull/100342#issuecomment-1212561869 seemed to say it's better off ignoring cases like `[u8; 64]` or similar.

"Basically just a lint looking for use of the mem_uninitialized diagnostic item" is what this was originally implemented as, and I can revert it back to that if we agree that's fine.

But **if** we wanted to reduce noise, then we could ignore things that are not UB (assuming the 0x01 filling or a hypothetical `freeze`).

Personally I think we've had MaybeUninit for long enough and that a full-on FCW for any mention of `mem::uninitialized` is warranted, but it *would* be noisy for some time while the ecosystem upgrades.

Reply:
Looking at it again, the guidance we got from the lang team [is ambiguous](https://github.com/rust-lang/rust/issues/98862#issuecomment-1209531725), so I asked for clarification in that thread.